abci/app.go

```diff
@@ -0,0 +1,378 @@
+	for _, v := range req.Validators {
+		app.updateValidator(v)
+	}
+	appHash := make([]byte, 8)
+	binary.PutVarint(appHash, app.state.Size)
+	req.ConsensusParams.Abci.VoteExtensionsEnableHeight = 1
+	return &abci.ResponseInitChain{ConsensusParams: req.ConsensusParams, AppHash: appHash}, nil
+}
+
+func (app *ForumApp) PrepareProposal(_ context.Context, proposal *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	fmt.Println("entered prepareProp")
+
+	voteExtensionCurseWords := app.getWordsFromVe(proposal.LocalLastCommit.Votes)
+
+	// prepare proposal puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range proposal.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err == nil {
+			if err != nil {
+				proposedTxs = append(proposedTxs, tx)
+			}
+			// Adding the curse words from vote extensions too
+			if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+				proposedTxs = append(proposedTxs, tx)
+			} else {
+				banTx := model.BanTx{UserName: msg.Sender}
+				bannedUsersString[msg.Message] = struct{}{}
+				resultBytes, err := json.Marshal(banTx)
+				if err == nil {
+					finalProposal = append(finalProposal, resultBytes)
+				}
+			}
+		}
+	}
+	// Need to loop again through the proposed Txs to make sure there is none left by a user that was banned after the tx was accepted
+	for _, tx := range proposedTxs {
+		// there should be no error here as these are just transactions we have checked and added
+		msg, _ := model.ParseMessage(tx)
+		if _, ok := bannedUsersString[msg.Sender]; !ok {
+			finalProposal = append(finalProposal, tx)
+		}
+	}
+	return &types.ResponsePrepareProposal{Txs: finalProposal}, nil
+}
+
+func (ForumApp) ProcessProposal(_ context.Context, processproposal *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+	fmt.Println("entered processProp")
+	processedBanTxs := false
+	bannedUsers := make(map[string]struct{}, 0)
+	for _, tx := range processproposal.Txs {
+		var parsedBan model.BanTx
+		var parsedTx model.Message
+		var err error
+		if !processedBanTxs && isBanTx(tx) {
+			fmt.Println("FoundBanTx")
+			err = json.Unmarshal(tx, &parsedBan)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, err
+			}
+			if processedBanTxs {
+				// Banning transactions have to come first, cannot have them once we hit the first non user ban tx
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, err
+			}
+			bannedUsers[parsedBan.UserName] = struct{}{}
+		} else {
+			_, err = model.ParseMessage(tx)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, err
+			}
+			processedBanTxs = true
```

Comment:
So, we're only able to process one ban tx?

Reply:
No, we will process as many ban transactions as they are. But they have to come one after the other - so before we hit this `else`. If they come after the `else` has been hit that means that the proposer did not put the ban transactions in front of the regular ones. (This is why in `PrepareProposal` banned transactions are appended directly to `finalProposal` and the regular ones are appended afterwards.